File: articles/narrative-identity-and-the-storied-self/python/analyze_narrative_identity.py
#!/usr/bin/env python3
from pathlib import Path
import json
import numpy as np
import pandas as pd

ROOT = Path(__file__).resolve().parents[1]
DATA = ROOT / "data" / "synthetic_narrative_identity.csv"
OUT = ROOT / "outputs"
OUT.mkdir(exist_ok=True)

THEMES = [
    "redemption",
    "contamination",
    "coherence",
    "agency",
    "communion",
    "meaning_making",
    "narrative_flexibility",
    "defensive_rigidity",
]
OUTCOMES = ["self_continuity", "well_being"]
DERIVED = [
    "narrative_growth_orientation",
    "narrative_burden",
    "narrative_integration",
    "redemptive_agency_balance",
]

def ols(df, y, xs):
    X = np.column_stack([np.ones(len(df)), df[xs].to_numpy(float)])
    beta, *_ = np.linalg.lstsq(X, df[y].to_numpy(float), rcond=None)
    return dict(zip(["intercept", *xs], map(float, beta)))

df = pd.read_csv(DATA)
required = {"person_id", "narrative_context", *THEMES, *OUTCOMES, *DERIVED}
missing = required - set(df.columns)
if missing:
    raise ValueError(f"Missing columns: {sorted(missing)}")

df["redemption_level"] = np.where(df["redemption"] > df["redemption"].median(), "higher_redemption", "lower_redemption")
df["contamination_level"] = np.where(df["contamination"] > df["contamination"].median(), "higher_contamination", "lower_contamination")
df["coherence_level"] = np.where(df["coherence"] > df["coherence"].median(), "higher_coherence", "lower_coherence")
df["narrative_profile"] = df["redemption_level"] + "_" + df["contamination_level"] + "_" + df["coherence_level"]

df["high_coherence_high_defensiveness"] = (
    (df["coherence"] > df["coherence"].median())
    & (df["defensive_rigidity"] > df["defensive_rigidity"].median())
)
df["high_contamination_low_continuity"] = (
    (df["contamination"] > df["contamination"].median())
    & (df["self_continuity"] < df["self_continuity"].median())
)

context_summary = (
    df.groupby("narrative_context")
    .agg(
        n=("person_id", "count"),
        redemption_mean=("redemption", "mean"),
        contamination_mean=("contamination", "mean"),
        coherence_mean=("coherence", "mean"),
        agency_mean=("agency", "mean"),
        communion_mean=("communion", "mean"),
        meaning_making_mean=("meaning_making", "mean"),
        narrative_flexibility_mean=("narrative_flexibility", "mean"),
        defensive_rigidity_mean=("defensive_rigidity", "mean"),
        self_continuity_mean=("self_continuity", "mean"),
        well_being_mean=("well_being", "mean"),
    )
    .reset_index()
)

profile_summary = (
    df.groupby("narrative_profile")
    .agg(
        n=("person_id", "count"),
        redemption_mean=("redemption", "mean"),
        contamination_mean=("contamination", "mean"),
        coherence_mean=("coherence", "mean"),
        agency_mean=("agency", "mean"),
        meaning_making_mean=("meaning_making", "mean"),
        self_continuity_mean=("self_continuity", "mean"),
        well_being_mean=("well_being", "mean"),
    )
    .reset_index()
)

pattern_summary = pd.DataFrame(
    {
        "pattern": ["high_coherence_high_defensiveness", "high_contamination_low_continuity"],
        "n": [int(df["high_coherence_high_defensiveness"].sum()), int(df["high_contamination_low_continuity"].sum())],
        "proportion": [float(df["high_coherence_high_defensiveness"].mean()), float(df["high_contamination_low_continuity"].mean())],
    }
)

corr_cols = THEMES + DERIVED + OUTCOMES
models = {
    "well_being": ols(df, "well_being", ["redemption", "contamination", "coherence", "agency", "communion", "meaning_making", "narrative_flexibility"]),
    "self_continuity": ols(df, "self_continuity", ["coherence", "redemption", "contamination", "meaning_making", "narrative_flexibility", "defensive_rigidity"]),
    "narrative_integration": ols(df, "narrative_integration", ["agency", "communion", "narrative_growth_orientation", "narrative_burden"]),
    "note": "Synthetic demonstration only; not clinical assessment, identity evaluation, personality testing, legal evaluation, workplace screening, or individual prediction.",
}

context_summary.to_csv(OUT / "python_narrative_context_summary.csv", index=False)
profile_summary.to_csv(OUT / "python_narrative_profile_summary.csv", index=False)
pattern_summary.to_csv(OUT / "python_narrative_pattern_summary.csv", index=False)
df[corr_cols].corr().to_csv(OUT / "python_narrative_identity_correlations.csv")
(OUT / "python_narrative_identity_models.json").write_text(json.dumps(models, indent=2))
df.to_csv(OUT / "python_scored_narrative_identity.csv", index=False)
print(f"Wrote outputs to {OUT}")

File: articles/narrative-identity-and-the-storied-self/r/analyze_narrative_identity.R
root <- normalizePath(file.path(dirname(sys.frame(1)$ofile), ".."))
data_path <- file.path(root, "data", "synthetic_narrative_identity.csv")
outputs <- file.path(root, "outputs")
dir.create(outputs, showWarnings = FALSE, recursive = TRUE)

data <- read.csv(data_path, stringsAsFactors = FALSE)

data$redemption_level <- ifelse(
  data$redemption > median(data$redemption),
  "higher_redemption",
  "lower_redemption"
)
data$contamination_level <- ifelse(
  data$contamination > median(data$contamination),
  "higher_contamination",
  "lower_contamination"
)
data$coherence_level <- ifelse(
  data$coherence > median(data$coherence),
  "higher_coherence",
  "lower_coherence"
)
data$narrative_profile <- paste(
  data$redemption_level,
  data$contamination_level,
  data$coherence_level,
  sep = "_"
)
data$high_coherence_high_defensiveness <- data$coherence > median(data$coherence) &
  data$defensive_rigidity > median(data$defensive_rigidity)
data$high_contamination_low_continuity <- data$contamination > median(data$contamination) &
  data$self_continuity < median(data$self_continuity)

context_summary <- aggregate(
  data[, c("redemption", "contamination", "coherence", "agency", "communion", "meaning_making", "narrative_flexibility", "defensive_rigidity", "self_continuity", "well_being")],
  by = list(narrative_context = data$narrative_context),
  FUN = mean
)
write.csv(context_summary, file.path(outputs, "r_narrative_context_summary.csv"), row.names = FALSE)

profile_summary <- aggregate(
  data[, c("redemption", "contamination", "coherence", "agency", "meaning_making", "self_continuity", "well_being")],
  by = list(narrative_profile = data$narrative_profile),
  FUN = mean
)
write.csv(profile_summary, file.path(outputs, "r_narrative_profile_summary.csv"), row.names = FALSE)

corr_cols <- c(
  "redemption", "contamination", "coherence", "agency", "communion",
  "meaning_making", "narrative_flexibility", "defensive_rigidity",
  "narrative_growth_orientation", "narrative_burden", "narrative_integration",
  "redemptive_agency_balance", "self_continuity", "well_being"
)
write.csv(cor(data[, corr_cols], use = "pairwise.complete.obs"), file.path(outputs, "r_narrative_identity_correlations.csv"))

model_wellbeing <- lm(
  well_being ~ redemption + contamination + coherence + agency +
    communion + meaning_making + narrative_flexibility,
  data = data
)

model_continuity <- lm(
  self_continuity ~ coherence + redemption + contamination +
    meaning_making + narrative_flexibility + defensive_rigidity,
  data = data
)

model_integration <- lm(
  narrative_integration ~ agency + communion +
    narrative_growth_orientation + narrative_burden,
  data = data
)

capture.output(summary(model_wellbeing), file = file.path(outputs, "r_wellbeing_model_summary.txt"))
capture.output(summary(model_continuity), file = file.path(outputs, "r_self_continuity_model_summary.txt"))
capture.output(summary(model_integration), file = file.path(outputs, "r_narrative_integration_model_summary.txt"))

write.csv(data, file.path(outputs, "r_scored_narrative_identity.csv"), row.names = FALSE)
cat("Wrote R outputs to:", outputs, "\n")
